### PR TITLE
Search cases

### PR DESCRIPTION
**Pull request by murilovmachado**
Jira: https://bugs.benetech.org/browse/CHI-282

This is the initial piece for search cases backend. The search query could no be built using sequelize, then we needed to implement it in plain SQL.

WIP:
- ~Filter by **helpline**~
- ~Filter by **phoneNumber** on households and perpetrators~
- ~Integration tests~

### CODE REVIEW

**Review by GPaoloni — COMMENTED** [review]
I'm not sure if I'm fully understanding all that's going on in the query, so the step by step explanation is welcome, but the idea looks good to me. Having the query in a separate file is a great approach. Just two things that you can improve:
- better indentation (imo, having the `case when then end` clause on a nested level between ANDs would help, just to give some example).
- [some comments](https://www.techonthenet.com/postgresql/comments.php) explaining at least the intention of the sub-queries may save our future selves some headaches.

File: controllers/case-controller.js [hunk 1]
```diff
@@ -75,7 +81,25 @@ const CaseController = Case => {
...
+    }
+    console.log({ query });
+    const [cases, metadata] = await sequelize.query(searchCasesQuery, {
+      replacements: {
+        helpline: body.helpline || null,
+        firstName: body.firstName ? `%${body.firstName}%` : null,
+        lastName: body.lastName ? `%${body.lastName}%` : null,
+        dateFrom: body.dateFrom || null,
+        dateTo: body.dateTo || null,
+        phoneNumber: body.phoneNumber ? `%${body.phoneNumber}%` : null,
```
Comment (nickhurlburt) [inline controllers/case-controller.js:96]: Do we need to strip non-digits out of the phone number here as in https://github.com/tech-matters/hrm/blob/master/controllers/contact-controller.js#L99 ?

Comment (nickhurlburt) [reply controllers/case-controller.js:96]: Also while I'm here, I'll mention to @GPaoloni that the hardcoded fields in https://github.com/tech-matters/hrm/blob/master/controllers/contact-controller.js#L111 may be a problem for customization.  Nothing to do on that for right now, just a thing to remember.

File: sql/search-cases-query.sql [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    households IS NOT NULL AND h.value->'household'->'name'->>'lastName' ILIKE :lastName
+    OR perpetrators IS NOT NULL AND p.value->'perpetrator'->'name'->>'lastName' ILIKE :lastName
+  )
+  END
+  AND CASE WHEN :dateFrom IS NULL THEN TRUE
+  ELSE cases."createdAt" >= :dateFrom::DATE
+  END
+  AND CASE WHEN :dateTo IS NULL THEN TRUE
+  ELSE cases."createdAt" <= :dateTo::DATE
+  END
```
Comment (nickhurlburt) [inline sql/search-cases-query.sql:36]: Just pinged Joan to see if there's anything more we want to do with date here (there may be some things in the future that we don't need to do now).

File: sql/search-cases-query.sql [hunk 3]
```diff
@@ -0,0 +1,53 @@
+SELECT DISTINCT ON (cases.id)
```
Comment (nickhurlburt) [inline sql/search-cases-query.sql:1]: I would agree with @GPaoloni's [comment](https://github.com/tech-matters/hrm/pull/54#pullrequestreview-5246420650) that we add some commenting and modify the indentation.  Though this is an epic and impressive query which taught me a few new Postgres tricks.  Nice.  Some integration tests may be a good idea too.

**Review by nickhurlburt — COMMENTED** [review]
Looks great, left a few comments.

**Pull request merged**